### PR TITLE
Do not filter out synthetic constructors that contain values of default arguments in Kotlin

### PR DESCRIPTION
**Pull request by Godin**
Similarly to #773 constructors in Kotlin can have default arguments with branches, what unfortunately was overlooked during implementation of #774 - currently we filter them out, because they gets compiled differently than functions:

```
class Example() {
    constructor(a: Boolean, b: String = if (a) "a" else "b") : this()
}
```

results in constructor with following signature

```
public synthetic <init>(ZLjava/lang/String;ILkotlin/jvm/internal/DefaultConstructorMarker;)V
```

This was discovered during implementation of #887.

### CODE REVIEW

File: org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/KotlinDefaultArgumentsFilter.java [hunk 1]
```diff
@@ -48,24 +48,38 @@ static boolean isDefaultArgumentsMethodName(final String methodName) {
...
 		}
-		if (!isDefaultArgumentsMethodName(methodNode.name)) {
-			return;
-		}
 		if (!KotlinGeneratedFilter.isKotlinClass(context)) {
 			return;
 		}
 
-		new Matcher().match(methodNode, output);
+		if (isDefaultArgumentsMethodName(methodNode.name)) {
```
Comment (marchof) [inline org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/KotlinDefaultArgumentsFilter.java]: For consistency with the new `isDefaultArgumentsConstructor()` method I would rename this method to `isDefaultArgumentsMethod()`. Also passing a `MethodNode` both all callers do `methodNode.name` anyways.

File: org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/SyntheticFilter.java [hunk 2]
```diff
@@ -35,6 +35,11 @@ public void filter(final MethodNode methodNode,
 				return;
 			}
 
+			if (KotlinDefaultArgumentsFilter
```
Comment (marchof) [inline org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/SyntheticFilter.java]: Side note: When I see this I more and more like @Godin 's idea to separate filters for different compilers.

Comment (Godin) [reply org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/SyntheticFilter.java]: @marchof don't know about whom you're talking 😆

Not sure that separation can remove this dependency between filters, i.e. not every but some `synthetic` methods anyway should be filtered in Kotlin.

File: org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/KotlinDefaultArgumentsFilter.java (hunk 1 above)
Comment (Godin) [reply org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/KotlinDefaultArgumentsFilter.java]: Done

File: org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/KotlinDefaultArgumentsFilter.java [hunk 3]
```diff
@@ -48,24 +48,38 @@ static boolean isDefaultArgumentsMethodName(final String methodName) {
 		return methodName.endsWith("$default");
 	}
 
+	static boolean isDefaultArgumentsConstructor(final MethodNode methodNode) {
+		if (!methodNode.name.equals("<init>")) {
```
[the remark this answers is not in the document]
Comment (Godin) [reply org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/KotlinDefaultArgumentsFilter.java]: I doubt that `methodNode.name` can be null, and `isDefaultArgumentsMethod` anyway performs `methodNode.name.endsWith` without null-check 😉 but flipped for consistency within `isDefaultArgumentsConstructor`.

**Comment by Godin** [comment]
@marchof fixed, also updated javadoc of filter according to changes, please re-review

**Pull request merged**